test/DGmethods/compressible_Navier_Stokes/mms_bc.jl

```diff
@@ -23,116 +23,115 @@ else
+vars_transform(::AtmosModel) = (:u, :v, :w)
+vars_diffusive(::AtmosModel) = (:τ11, :τ22, :τ33, :τ12, :τ13, :τ23)
+
+function flux!(::AtmosModel, flux::Grad, state::State, diffstate::State, auxstate::State, t::Real)
+  # preflux
+  γ = γ_exact  
+  ρinv = 1 / state.ρ
+  u, v, w = ρinv * state.ρu, ρinv * state.ρv, ρinv * state.ρw
+  P = (γ-1)*(state.ρe - ρinv * (state.ρu^2 + state.ρv^2 + state.ρw^2) / 2)
+
+  # invisc terms
+  flux.ρ  = (state.ρu          , state.ρv          , state.ρw)
+  flux.ρu = (u * state.ρu  + P , v * state.ρu      , w * state.ρu)
+  flux.ρv = (u * state.ρv      , v * state.ρv + P  , w * state.ρv)
+  flux.ρw = (u * state.ρw      , v * state.ρw      , w * state.ρw + P)
+  flux.ρe = (u * (state.ρe + P), v * (state.ρe + P), w * (state.ρe + P))
+
+  # viscous terms
+  flux.ρu .-= (diffstate.τ11, diffstate.τ12, diffstate.τ13)
+  flux.ρv .-= (diffstate.τ12, diffstate.τ22, diffstate.τ23)
+  flux.ρw .-= (diffstate.τ13, diffstate.τ23, diffstate.τ33)
+
+  flux.ρe .-= (u * diffstate.τ11 + v * diffstate.τ12 + w * diffstate.τ13,
+               u * diffstate.τ12 + v * diffstate.τ22 + w * diffstate.τ23,
+               u * diffstate.τ13 + v * diffstate.τ23 + w * diffstate.τ33)
 end
 
-include("mms_solution_generated.jl")
+function transform!(::AtmosModel, transformstate::State, state::State, auxstate::State, t::Real)
+  ρinv = 1 / state.ρ
+  transformstate.u = ρinv * state.ρu,
+  transformstate.v = ρinv * state.ρv
+  transformstate.w = ρinv * state.ρw
+end
 
-# preflux computation
-@inline function preflux(Q, _...)
-  γ::eltype(Q) = γ_exact
-  @inbounds ρ, U, V, W, E = Q[_ρ], Q[_U], Q[_V], Q[_W], Q[_E]
-  ρinv = 1 / ρ
-  u, v, w = ρinv * U, ρinv * V, ρinv * W
-  ((γ-1)*(E - ρinv * (U^2 + V^2 + W^2) / 2), u, v, w, ρinv)
+function diffusive!(::AtmosModel, diff::State, ∇transform::Grad)
+  dudx, dudy, dudz = ∇transform.u
+  dvdx, dvdy, dvdz = ∇transform.v
+  dwdx, dwdy, dwdz = ∇transform.w
+
+  # strains
+  ϵ11 = dudx
+  ϵ22 = dvdy
+  ϵ33 = dwdz
+  ϵ12 = (dudy + dvdx) / 2
+  ϵ13 = (dudz + dwdx) / 2
+  ϵ23 = (dvdz + dwdy) / 2
+
+  # deviatoric stresses
+  diff.τ11 = 2μ * (ϵ11 - (ϵ11 + ϵ22 + ϵ33) / 3)
+  diff.τ22 = 2μ * (ϵ22 - (ϵ11 + ϵ22 + ϵ33) / 3)
+  diff.τ33 = 2μ * (ϵ33 - (ϵ11 + ϵ22 + ϵ33) / 3)
+  diff.τ12 = 2μ * ϵ12
+  diff.τ13 = 2μ * ϵ13
+  diff.τ23 = 2μ * ϵ23
 end
 
-# max eigenvalue
-@inline function wavespeed(n, Q, aux, t, P, u, v, w, ρinv)
-  γ::eltype(Q) = γ_exact
-  @inbounds abs(n[1] * u + n[2] * v + n[3] * w) + sqrt(ρinv * γ * P)
+function source!(::AtmosModel, source::State, state::State, aux::State, t::Real)
+  S.ρ  = Sρ_g(t, aux.x, aux.y, aux.z, Val(3))
+  S.ρu = SU_g(t, aux.x, aux.y, aux.z, Val(3))
+  S.ρv = SV_g(t, aux.x, aux.y, aux.z, Val(3))
+  S.ρw = SW_g(t, aux.x, aux.y, aux.z, Val(3))
+  S.ρe = SE_g(t, aux.x, aux.y, aux.z, Val(3))
 end
 
-# flux function
-cns_flux!(F, Q, VF, aux, t) = cns_flux!(F, Q, VF, aux, t, preflux(Q)...)
+function wavespeed(::AtmosModel, nM, state::State, aux::State, t::Real)
```

Comment:
This isn't general (not correct in moist case). Use `soundspeed_air` in `MoistThermodynamics` instead. To make the `DGmethods` entirely general, it would be nice to use a pointer for this function that just for the case of an atmosphere points to `suundspeed_air` (and, e.g., for the ocean, to its equivalent in the ocean). Same for `air_pressure` above (pointer to generic pressure function).